src/transformStateWithClones.js

```diff
@@ -6,8 +6,46 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStates = [];
+  const realState = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(realState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(realState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(realState);
+        break;
+    }
```

Comment:
[CODE STYLE]: The `switch` statement should include a `default` case to handle unexpected `action.type` values. This is important for error handling and to ensure that the function behaves correctly even if it encounters an unknown action.